Fix bug where bundleMapQueue is not cleared after an error

morwoen opened the pull request:
Fixes #69 

# What happened
At the Localisation & Globalisation team at my company, we've decided to release an internal "test" locale (we called it xx-XX) to provide pseudo localisation for all other teams across. Since this locale is fictional there is no CLDR data for it. Our approach was to take es-ES and alter the data so that it acts as the new xx-XX locale.

During the process we've missed the likelySubtags object, meaning there was a mismatch
```
"main": {
  "xx-XX": {...}
},
"supplemental": {
  "likelySubtags": {
    "und": "es",
    "es-ES": "es"
  }
}
```

# The error
Passing this data into a `Globalize.load()` (which in turn calls `CLDR.load()`) causes no errors, however, instantiating a new instance `new Globalize(locale)` throws `TypeError: Cannot read property '0' of undefined`. Note that what `locale` is above doesn't matter. Locales with already loaded valid data can no longer be used.

# Expected outcome
An error to be thrown *only once* and the data is ejected from the queue to allow the application to continue without a need to restart.

# Impact
The service that observed the error is using Globalize in its backend. This means that after an internal user (xx-XX is only an internal locale, no end-user can request it) tested the new locale end-users using actual locales started seeing errors due to the recurring exception thrown for valid locales.

# Why this fixed it
I have moved the clearing of the _availableBundleMapQueue to before the iteration over it. That way even if there is an error in the code below the error will happen only once. There appear to be no other exit points before the queue clearing, so pulling it earlier is not changing the behaviour.

morwoen committed:
Fix bug where bundleMapQueue is not cleared after an error

Signed-off-by: Georgi Tsaklev <georgi.tsaklev@skyscanner.net>

morwoen committed:
Fix linter errors

Signed-off-by: Georgi Tsaklev <georgi.tsaklev@skyscanner.net>

rxaviers commented on test/unit/bundle/lookup.js:
```diff
@@ -55,4 +55,23 @@ describe("Bundle Lookup", function() {
+      expect(Cldr._availableBundleMapQueue).to.include.members(["xx-XX"]);
+      new Cldr("xx-XX");
+    }).to.throw();
+    expect(Cldr._availableBundleMapQueue).to.be.empty;
```
I'm wondering if we could do this:

```js
expect(() => {
  Cldr.load(
    {
      main: { "xx-XX": {} }
    },
    {
      main: { sr: {} }
    }
  );
  expect(Cldr._availableBundleMapQueue).to.include.members(["xx-XX", "sr"]);
  new Cldr("xx-XX");
}).to.throw();
expect(Cldr._availableBundleMapQueue).to.include.members(["sr"]);
```

morwoen replied:
Thanks for the review :)

On a higher level, what would you expect to happen? If we create an instance of sr instead of xx-XX it would still throw because it would iterate in order. The test as it is would succeed and a second attempt of creating an instance of sr would succeed with the data parsed. Does this sound like an acceptable behaviour to you?

rxaviers replied:
> On a higher level, what would you expect to happen? If we create an instance of sr instead of xx-XX it would still throw because it would iterate in order

That's a good question. Yeap, I'd expect that to still throw. That being said, ideally we have a better error message saying it's missing supplemental.likelySubtags data for xx-XX.

rxaviers replied:
> The test as it is would succeed and a second attempt of creating an instance of sr would succeed with the data parsed. Does this sound like an acceptable behaviour to you?

Yes, though I am still not convinced about the solution. Do we really need this code change, or all we need is a better error message (comment above) and simply expect Cldrjs user to feed it on the right data?

The root cause of this issue is the artificial `{main: {"xx-XX": {...}}` without a proper likelySubtags. Similar issue could happen for CLDR data if one loads main without loading likelySubtags. Ideally, the root cause is fixed, i.e., user (i.e., engineer/developer) should pass the right data to CLDR.load.

What do you think?

morwoen replied:
I agree with the root cause, however developers make mistakes and bad data would be passed sooner or later. I would expect to still be able to use the library for other data that is already loaded and newly loaded valid data rather than getting an error on anything after the invalid one is loaded. Currently the whole service needs to be recycled before it can serve requests again.

I am all for adding a more descriptive error message. I can manually throw a descriptive message between line 24 and 25 if coreLikelySubtags returns undefined which would help future debugging.

I guess we can have two approaches here -
1. load all data that can be loaded, collect all errors and throw at the end after everything has been processed
2. load everything that can be loaded until an error is observed. In this case stop iterating and remove everything from the queue including the failed bundle

rxaviers replied:
So let's try-catch that error, add a more descriptive error message and splice `Cldr._availableBundleMapQueue` to remove the failing item. What do you think?

morwoen replied:
Yeah, happy with this approach. And just to confirm this means it will stop processing bundles on error, but next time it will succeed as the erroneous bundle will be removed and it will start from the one after?

rxaviers replied:
Exactly this. Thanks

rxaviers commented:
Thank you for reporting the issue and for providing a solution in this PR. I have left one comment above. Thanks

morwoen committed:
Updated lookup logic for processing _availableBundleMapQueue to use .shift instead of .forEach

Signed-off-by: Georgi Tsaklev <georgi.tsaklev@skyscanner.net>

morwoen commented:
Hi @rxaviers ,
I've updated the PR around what we discussed. It now mutates the array as it iterates over it so that when an error is thrown the problematic bundle will already be removed from the queue. I've added a more descriptive error message, feel free to suggest improvement to the language used.

I've also added some more comments to the test so it is easier to understand in the future for new people and once we forget about this :)

rxaviers commented:
Awesome! Looks great. Thanks a lot

rxaviers commented on src/bundle/lookup.js:
```diff
@@ -17,19 +16,27 @@ export default function(Cldr, cldr, minLanguageId) {
       minBundle = minBundle.join(Cldr.localeSep);
       existing = availableBundleMap[minBundle];
       if (existing && existing.length < bundle.length) {
         return;
```
Late review... This should also be `break`.

rxaviers commented:
Amended by 2bac803c1ebe234f4abc22b38b314c65523a63d9

PS: In addition to the s/return/break mentioned above, there's something else that was causing regressions in PayPal libs I couldn't debug (time-wise). Amended using code closest to what we had before while keeping what @morwoen needed (please let me know otherwise). Thx

morwoen commented:
Ah, Completely missed that return :) yeah, I think it still achieves what we aimed for, thanks for validating it with your internal libs 🥇
